client/json-rpc/src/client.rs

```diff
@@ -115,6 +115,33 @@ impl JsonRpcBatch {
     }
 }
 
+#[derive(Debug)]
+pub enum JsonRpcAsyncClientError {
+    SubmissionError(reqwest::Error),
```

Comment:
can we name this ClientError  / InvalidServerResponse to match what the python library is doing?


cc @ssinghaldev  to provide more context around this

Reply:
Tbh we're less interested about Client vs. Server dichotomy, and more about whether to retry or not. For example, we don't want to retry in case a given transaction cannot be properly serialized (but it's not a `InvalidServerResponse`). In that sense, Execution vs Submission could be a nice way to partition things. 
No strong opinions about naming though. We could also just add more enum options to be more specific about the error types.

Reply:
After reading the code I'd like to propose a more detailed enum with more values, e.g., `WrongInput`, `HttpError` (with http error code), `ConnectivityError`. The client could then determine the retry policy for themselves.

Reply:
@dmitri-perelman I think the reqwest::Error already includes all three types of error you mentioned already. I'm not against having a more detailed enum though, because one risk we are taking here is that we are making ourselves depend on specifics of reqwest here, may make it hard to replace in the future

Reply:
> can we name this ClientError / InvalidServerResponse to match what the python library is doing?
> 
> cc @ssinghaldev to provide more context around this

Just for adding info on how we are doing it in python client. Errors are ValueError(WrongInput), RequestException(ConnectionError), InvalidServerResponse(deserializing & response is not according to json-rpc spec), JsonRpcError(which is when the rpc returns "error" instead of result). These errors are from json-rpc client I have written. But the actual API (which uses json-rpc underneath) throws ValueError(re-raises basically), packages RequestException, InvalidServerResponse, JsonRpcError  into ClientError & throws that. For now, the idea is ValueError represents - u can't retry with same options but with ClientError u can. Ofccourse, it can be more fine grained. But for now, we have this. 

The idea of repackaging is basically @ZhenjieRuan mentioned that we do not wish to depend/pass on Library error(in my context, it is request & underlying json-rpc client which I have written)

Reply:
Sure, no objection for renaming.